### PR TITLE
Admin Support User and Event API

### PR DESCRIPTION
**Pull request by Krace33**

### CODE REVIEW

File: src/main/java/com/pecacm/backend/services/SupportService.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+@Service
+public class SupportService {
+    private final TransactionRepository transactionRepository;
+
+    public SupportService(TransactionRepository transactionRepository) {
+        this.transactionRepository = transactionRepository;
+    }
+
+    public SupportUserResponse getSupportUserDetails(User user, PageRequest pageRequest) {
+        Page<Transaction> transactions = transactionRepository.findByUserId(user.getId(), pageRequest);
```
Comment (harshjohar) [inline src/main/java/com/pecacm/backend/services/SupportService.java:30]: You can edit this query to get the transactions via email of the user by joining the two tables (users and transactions)

File: src/main/java/com/pecacm/backend/services/SupportService.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        List<UserEventDetails> events = new ArrayList<>();
+        transactions.forEach(transaction -> {
+            Event event = transaction.getEvent();
+            UserEventDetails userEventDetails = new UserEventDetails(event.getId(), event.getTitle(), transaction.getRole(), transaction.getXp(), event.getEndDate());
+            events.add(userEventDetails);
+        });
```
Comment (harshjohar) [inline src/main/java/com/pecacm/backend/services/SupportService.java:36]: A better way of doing this is using java streams and collectors
```suggestion
        List<UserEventDetails> events = transactions.stream()
    .map(transaction -> {
        Event event = transaction.getEvent();
        return new UserEventDetails(event.getId(), event.getTitle(), transaction.getRole(), transaction.getXp(), event.getEndDate());
    })
    .collect(Collectors.toList());
```

File: src/main/java/com/pecacm/backend/response/SupportUserResponse.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.Setter;
+import org.springframework.data.domain.Page;
+
+
+@Getter
+@Setter
+@AllArgsConstructor
+public class SupportUserResponse {
+    private User user;
+    private Page<UserEventDetails> events;
```
Comment (harshjohar) [inline src/main/java/com/pecacm/backend/response/SupportUserResponse.java:15]: I dont like this Page in response, but will fix this in some common class in later issues

**Review by harshjohar — CHANGES_REQUESTED** [review]

File: src/main/java/com/pecacm/backend/controllers/SupportController.java [hunk 4]
```diff
@@ -1,10 +1,69 @@
...
+    public ResponseEntity<SupportUserResponse>  getUserDetails(@PathVariable String email, @RequestParam @Nullable Integer offset, @RequestParam @Nullable Integer pageSize) {
+
+        if (offset == null) offset = 0;
+        if (pageSize == null) pageSize = 20;
+
+        if (offset < 0) throw new AcmException("offset cannot be < 0", HttpStatus.BAD_REQUEST);
+        if (pageSize <= 0) throw new AcmException("pageSize must be >= 0", HttpStatus.BAD_REQUEST);
+
+        User user = userService.getUserByEmail(email);
+        return ResponseEntity.ok(supportService.getSupportUserDetails(user, PageRequest.of(offset, pageSize)));
```
[the remark this answers is not in the document]
Comment (Krace33) [reply src/main/java/com/pecacm/backend/controllers/SupportController.java:50]: User object needed for response anyway

File: src/main/java/com/pecacm/backend/services/SupportService.java (hunk 1 above)
Comment (Krace33) [reply src/main/java/com/pecacm/backend/services/SupportService.java:30]: User object needed for response anyway

File: src/main/java/com/pecacm/backend/services/SupportService.java (hunk 2 above)
Comment (Krace33) [reply src/main/java/com/pecacm/backend/services/SupportService.java]: Replaced redundant for loops with streams

File: src/main/java/com/pecacm/backend/controllers/SupportController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Krace33) [reply src/main/java/com/pecacm/backend/controllers/SupportController.java]: Done

**Review by harshjohar — APPROVED** [review]
LGTM

File: src/main/java/com/pecacm/backend/response/UserEventDetails.java
Comment (harasees-singh) [inline src/main/java/com/pecacm/backend/response/UserEventDetails.java:1]: please change it to 'UserEventResponse.java' with the following code 

```package com.pecacm.backend.response;

import com.pecacm.backend.enums.EventRole;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;

@AllArgsConstructor
@Setter
@Getter
public class UserEventResponse {
    private Integer eventId;
    private String eventName;
    private EventRole eventRole;
    private Integer xp_gained;
    private LocalDateTime eventEnd;
}```

File: src/main/java/com/pecacm/backend/services/SupportService.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    public SupportUserResponse getSupportUserDetails(String email, PageRequest pageRequest) {
+        User user = userService.getUserByEmail(email);
+        Page<Transaction> transactions = transactionRepository.findByUserId(user.getId(), pageRequest);
+        List<UserEventDetails> events = transactions.stream()
+                .map(transaction -> {
+                    Event event = transaction.getEvent();
+                    return new UserEventDetails(event.getId(), event.getTitle(), transaction.getRole(), transaction.getXp(), event.getEndDate());
+                }).collect(Collectors.toList());
```
Comment (harasees-singh) [inline src/main/java/com/pecacm/backend/services/SupportService.java:39]: nit: can simplify .collect(Collectors.toList()) to .toList()

**Pull request merged**